### PR TITLE
[Doc] Add dataflow document

### PR DESCRIPTION
**Pull request by xiexinch**
## Motivation

Add dataflow doc

## Modification

- docs/en/advanced_guides/data_flow.md

### CODE REVIEW

File: docs/en/advanced_guides/data_flow.md [hunk 1]
```diff
@@ -1 +1,57 @@
...
+# Dataflow
+
+In this chapter, we will introduce the dataflow and data format convention between the internal modules managed by the [Runner](https://mmengine.readthedocs.io/en/latest/tutorials/runner.html).
+
+## Overview of dataflow
+
+As illustrated in the [Runner document of MMEngine](https://mmengine.readthedocs.io/en/latest/tutorials/runner.html), the following diagram shows the basic dataflow.
+
+![Basic dataflow](https://user-images.githubusercontent.com/112053249/199228350-5f80699e-7fd2-4b4c-ac32-0b16b1922c2e.png)
+
```
Comment (MengzhangLI) [inline docs/en/advanced_guides/data_flow.md:10]: Add detail explanation for red/blue/green lines. Just like what [MMEngine](https://github.com/open-mmlab/mmengine/blob/main/docs/en/tutorials/runner.md) did.

image

File: docs/en/advanced_guides/data_flow.md [hunk 2]
```diff
@@ -1 +1,57 @@
...
+The dashed border, gray filled shapes represent different data formats, while solid boxes represent modules/methods. Due to the great flexibility and extensibility of MMEngine, you can always inherit some key base classes and override their methods, so the above diagram doesn’t always hold. It only holds when you are not customizing your own `Runner` or `TrainLoop`, and you are not overriding `train_step`, `val_step` or `test_step` method in your custom model.
```
Comment (MengzhangLI) [inline docs/en/advanced_guides/data_flow.md:11]:
```suggestion
The dashed border, gray filled shapes represent different data formats, while solid boxes represent modules/methods. Due to the great flexibility and extensibility of MMEngine, some critical base classes can be inherited and and their methods can be overridden. The diagram above only holds when users are not customizing `TrainLoop`, `ValLoop`, and `TestLoop` in `Runner`, and are not overriding `train_step`, `val_step` and `test_step` method in their custom model.
```
If customization of dataflow is not very important, maybe we should delete the last sentence.

File: docs/en/advanced_guides/data_flow.md [hunk 3]
```diff
@@ -1 +1,57 @@
...
+
+## Format convention
+
```
Comment (MengzhangLI) [inline docs/en/advanced_guides/data_flow.md:14]: From the diagram above, we could see the basic dataflow. In this section, we would introduce format convention of data involved in this dataflow, respectively.

File: docs/en/advanced_guides/data_flow.md [hunk 4]
```diff
@@ -1 +1,57 @@
...
+### DataLoader to DataPreprocessor
```
Comment (MengzhangLI) [inline docs/en/advanced_guides/data_flow.md:15]:
```suggestion
### From DataLoader to Data Preprocessor
```
To keep the same with next subsection `Data Preprocessor`.

File: docs/en/advanced_guides/data_flow.md [hunk 5]
```diff
@@ -1 +1,57 @@
...
+
+DataLoader is a necessary component in MMEngine's training and testing pipelines. It's conceptually derived from and consistent with PyTorch. DataLoader loads data from filesystem and the original data pass through data prepare pipeline then would be sent to the DataPreprocessor. MMSegmentation defines the default data format at [PackSegInputs](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/datasets/transforms/formatting.py#L12), it's the last component of `train_pipeline` and `test_pipeline`.
```
Comment (MengzhangLI) [inline docs/en/advanced_guides/data_flow.md:17]:
```suggestion
DataLoader is an essential component in training and testing pipelines of MMEngine. Conceptually, it is derived from and consistent with PyTorch (add pytorch official ref link). DataLoader loads data from filesystem and the original data passes through data preparation pipeline, then it would be sent to Data Preprocessor. 

MMSegmentation defines the default data format at [PackSegInputs](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/datasets/transforms/formatting.py#L12), it's the last component of `train_pipeline` and `test_pipeline`. 
```
Maybe add a short explanation and link about `pipeline`.

File: docs/en/advanced_guides/data_flow.md [hunk 6]
```diff
@@ -1 +1,57 @@
...
+    inputs=torch.Tensor,
+    data_samples=SegDataSample
+)
+```
+
+**Note:** [SegDataSample](https://github.com/open-mmlab/mmsegmentation/blob/1.x/mmseg/structures/seg_data_sample.py) is a data structure interface of MMSegmentation, it is used as an interface between different components. `SegDataSample` implements the abstract data element `mmengine.structures.BaseDataElement`, please refer to [the SegDataSample documentation](https://mmsegmentation.readthedocs.io/en/1.x/advanced_guides/structures.html) and [data element documentation](https://mmengine.readthedocs.io/en/latest/advanced_tutorials/data_element.html) in [MMEngine](https://github.com/open-mmlab/mmengine) for more information.
+
+### Data Preprocessor to Model
+
+Though drawn separately in the diagram [above](#overview-of-dataflow), data_preprocessor is a part of the model and thus can be found in [Model tutorial](./models.md) at Seg DataPreprocessor chapter.
```
Comment (MengzhangLI) [inline docs/en/advanced_guides/data_flow.md:32]: Add specific http link of `models.md`.

File: docs/en/advanced_guides/data_flow.md [hunk 7]
```diff
@@ -1 +1,57 @@
...
+
+The return value is the same as `PackSegInputs` except the `inputs` would be transferred to GPU and some additional metainfo like `pad_shape` and `padding_size` would be added to the `data_samples`.
```
Comment (MengzhangLI) [inline docs/en/advanced_guides/data_flow.md:34]:
```suggestion
In the return value of Data Preprocessor, `inputs` would be transferred to GPU and some additional metainfo like `pad_shape` and `padding_size` would be added to the `data_samples`.
```
The readers may be confused when reading "the same as `PackSegInputs` except ..." because the differences are what data preprocessor do in data flow.

File: docs/en/advanced_guides/data_flow.md [hunk 8]
```diff
@@ -1 +1,57 @@
...
+
+### Model to Evaluator
```
Comment (MengzhangLI) [inline docs/en/advanced_guides/data_flow.md:36]:
```suggestion
### Model to Evaluator in val/test step
```

File: docs/en/advanced_guides/data_flow.md [hunk 9]
```diff
@@ -1 +1,57 @@
...
+
+At the evaluation procedure, the inference results would be transferred to `Evaluator`. You might read the [evaluation document](./evaluation.md) for more information about `Evaluator`.
```
Comment (MengzhangLI) [inline docs/en/advanced_guides/data_flow.md:38]: Add specific http link of `evaluation.md`.

File: docs/en/advanced_guides/data_flow.md [hunk 10]
```diff
@@ -1 +1,57 @@
...
+
+After inference, the [BaseSegmentor](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/segmentors/base.py#L15) in MMSegmentation would do a simple post process to pack inference resutls, the segmentation logits produced by the neural network, segmentation mask after the `argmax` operation and ground truth(if exists) would be pack into a same SegDataSample instance. The return value of [postprocess_result](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/segmentors/base.py#L132) is a **`List` of `SegDataSample`**. Following diagram shows the key properties of these SegDataSample instances.
+
+![SegDataSample](../../../resources/SegDataSample.png)
+
+### Model to Loss function
```
Comment (MengzhangLI) [inline docs/en/advanced_guides/data_flow.md:44]:
```suggestion
### Model to Loss function in train step
```

File: docs/en/advanced_guides/data_flow.md [hunk 11]
```diff
@@ -1 +1,57 @@
...
+
+The return value is the same as `PackSegInputs` except the `inputs` would be transferred to GPU and some additional metainfo like `pad_shape` and `padding_size` would be added to the `data_samples`.
+
+### Model to Evaluator
+
+At the evaluation procedure, the inference results would be transferred to `Evaluator`. You might read the [evaluation document](./evaluation.md) for more information about `Evaluator`.
+
+After inference, the [BaseSegmentor](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/segmentors/base.py#L15) in MMSegmentation would do a simple post process to pack inference resutls, the segmentation logits produced by the neural network, segmentation mask after the `argmax` operation and ground truth(if exists) would be pack into a same SegDataSample instance. The return value of [postprocess_result](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/segmentors/base.py#L132) is a **`List` of `SegDataSample`**. Following diagram shows the key properties of these SegDataSample instances.
```
Comment (MengzhangLI) [inline docs/en/advanced_guides/data_flow.md:40]:
```suggestion
After inference, the [BaseSegmentor](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/segmentors/base.py#L15) in MMSegmentation would do a simple post process to pack inference results, the segmentation logits produced by the neural network, segmentation mask after the `argmax` operation and ground truth (if exists) would be packed into a similar `SegDataSample` instance. The return value of [postprocess_result](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/segmentors/base.py#L132) is a **`List` of `SegDataSample`**. Following diagram shows the key properties of these `SegDataSample` instances.
```

File: docs/en/advanced_guides/data_flow.md [hunk 12]
```diff
@@ -1 +1,57 @@
...
+
+![SegDataSample](../../../resources/SegDataSample.png)
+
+### Model to Loss function
+
+The same as Data Preprocessor, loss function is also a part of the model, it's a property of [decode head](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/decode_heads/decode_head.py#L142).
+
+In MMSegmentation, the method [loss_by_feat](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/decode_heads/decode_head.py#L291) of `decode_head` is a unify interface used to compute loss.
```
Comment (MengzhangLI) [inline docs/en/advanced_guides/data_flow.md:48]:
```suggestion
In MMSegmentation, the method [loss_by_feat](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/decode_heads/decode_head.py#L291) of `decode_head` is an unified interface used to compute loss.
```

File: docs/en/advanced_guides/data_flow.md (hunk 8 above)
Comment (MeowZheng) [inline docs/en/advanced_guides/data_flow.md:36]: we can add some explanation about  the control flow of runner when training and testing, like call train_step: model.forward-> evaluator, call test_step: model.forward -> optim wrapper refer, and attach the link https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/docs/en/advanced_guides/models.md

and highlight the input and output of model and other modules

File: docs/en/advanced_guides/data_flow.md [hunk 13]
```diff
@@ -1 +1,85 @@
...
+
+### Model output
+
+#### To Evaluator
+
+At the evaluation procedure, the inference results would be transferred to `Evaluator`. You might read the [evaluation document](https://mmsegmentation.readthedocs.io/en/dev-1.x/advanced_guides/evaluation.html) for more information about `Evaluator`.
+
+After inference, the [BaseSegmentor](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/segmentors/base.py#L15) in MMSegmentation would do a simple post process to pack inference results, the segmentation logits produced by the neural network, segmentation mask after the `argmax` operation and ground truth(if exists) would be packed into a similar `SegDataSample` instance. The return value of [postprocess_result](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/segmentors/base.py#L132) is a **`List` of `SegDataSample`**. Following diagram shows the key properties of these `SegDataSample` instances.
+
+![SegDataSample](../../../resources/SegDataSample.png)
```
Comment (MeowZheng) [inline docs/en/advanced_guides/data_flow.md:70]: might also use github picture-bed

File: docs/en/advanced_guides/data_flow.md [hunk 14]
```diff
@@ -1 +1,85 @@
...
+train_cfg = dict(type='IterBasedTrainLoop', max_iters=20000, val_interval=2000)
+val_cfg = dict(type='ValLoop')
+test_cfg = dict(type='TestLoop')
+```
+
+In the above diagram, the red line indicates the [train_step](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/docs/en/advanced_guides/models.md#train_step). At each training iteration, dataloader loads images from storage and transfer to data preprocessor, data preprocessor would put images to GPU device and stack data to batch, then model accept the batch data as inputs, finally the outputs of the model would be sent to optimizer. The blue line indicates [val_step](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/docs/en/advanced_guides/models.md#val_step) and [test_step](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/docs/en/advanced_guides/models.md#test_step). The dataflow of these two process is similar to the `train_step` except the outputs of model, since model parameters are freezed when doing evaluation, the model output would be transferred to [Evaluator](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/docs/en/advanced_guides/evaluation.md#ioumetric) to compute metrics.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/data_flow.md:19]:
```suggestion
In the above diagram, the red line indicates the [train_step](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/docs/en/advanced_guides/models.md#train_step). At each training iteration, dataloader loads images from storage and transfer to data preprocessor, data preprocessor would put images to the specific device and stack data to batch, then model accepts the batch data as inputs, finally the outputs of the model would be sent to optimizer. The blue line indicates [val_step](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/docs/en/advanced_guides/models.md#val_step) and [test_step](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/docs/en/advanced_guides/models.md#test_step). The dataflow of these two process is similar to the `train_step` except the outputs of model, since model parameters are freezed when doing evaluation, the model output would be transferred to [Evaluator](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/docs/en/advanced_guides/evaluation.md#ioumetric) to compute metrics.
```

File: docs/en/advanced_guides/data_flow.md [hunk 15]
```diff
@@ -1 +1,85 @@
...
+```python
+class Network(BaseSegmentor):
+
+    def forward(self, inputs: torch.Tensor, data_samples: List[SegDataSample], mode: str):
+        pass
+```
+
+### Model output
+
+#### To Evaluator
```
Comment (MeowZheng) [inline docs/en/advanced_guides/data_flow.md:64]:
```suggestion
As [model tutorial](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/docs/en/advanced_guides/models.md#forward) mentioned 3 kinds of mode forward with 3 kinds of output. `train_step`and `test_step`(or `val_step`) correspond to `'loss'` and `'predict'` respectively.
```

File: docs/en/advanced_guides/data_flow.md [hunk 16]
```diff
@@ -1 +1,85 @@
...
+
+At the evaluation procedure, the inference results would be transferred to `Evaluator`. You might read the [evaluation document](https://mmsegmentation.readthedocs.io/en/dev-1.x/advanced_guides/evaluation.html) for more information about `Evaluator`.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/data_flow.md:66]:
```suggestion
In `test_step` or `val_step`, the inference results would be transferred to `Evaluator`. You might read the [evaluation document](https://mmsegmentation.readthedocs.io/en/dev-1.x/advanced_guides/evaluation.html) for more information about `Evaluator`.
```

File: docs/en/advanced_guides/data_flow.md [hunk 17]
```diff
@@ -1 +1,85 @@
...
+
+After inference, the [BaseSegmentor](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/segmentors/base.py#L15) in MMSegmentation would do a simple post process to pack inference results, the segmentation logits produced by the neural network, segmentation mask after the `argmax` operation and ground truth(if exists) would be packed into a similar `SegDataSample` instance. The return value of [postprocess_result](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/segmentors/base.py#L132) is a **`List` of `SegDataSample`**. Following diagram shows the key properties of these `SegDataSample` instances.
+
+![SegDataSample](../../../resources/SegDataSample.png)
+
+#### Optim Wrapper
```
Comment (MeowZheng) [inline docs/en/advanced_guides/data_flow.md:72]:
```suggestion
```

File: docs/en/advanced_guides/data_flow.md [hunk 18]
```diff
@@ -1 +1,85 @@
...
+In MMSegmentation, the method [loss_by_feat](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/mmseg/models/decode_heads/decode_head.py#L291) of `decode_head` is an unified interface used to compute loss.
+
+Parameters:
+
+- seg_logits (Tensor): The output from decode head forward function.
+- batch_data_samples (List\[:obj:`SegDataSample`\]): The seg data samples. It usually includes information such as `metainfo` and `gt_sem_seg`.
+
+Returns:
+
+- dict\[str, Tensor\]: a dictionary of loss components
```
Comment (MeowZheng) [inline docs/en/advanced_guides/data_flow.md:85]: The `train_step` transfers the loss into OptimWrapper to update the weights in model, please refer [`train_step`](https://github.com/open-mmlab/mmengine/blob/6af88783fb0c7556cfcdd09c7ad839f9572c3095/mmengine/model/base_model/base_model.py#L84) for more details.

**Review by MeowZheng — APPROVED** [review]
